edisgo/tools/tools.py

```diff
@@ -193,7 +193,109 @@ def drop_duplicated_columns(df, keep="last"):
+) -> float | np.ndarray:
+    """
+    Calculate the voltage difference across a line in kV.
+
+    Parameters
+    ----------
+    s_max : float or array-like
+        Apparent power the cable must carry in MVA.
+    r_total : float or array-like
+        Total resistance in Ohms.
+    x_total : float or array-like
+        Total reactance in Ohms.
+    v_nom : float or array-like
+        Nominal voltage in kV.
+    sign : int, optional
+        Sign of the reactance. -1 for inductive and +1 for capacitive. Default is -1.
+    cos_phi : float, optional
+        Power factor (cosine of the phase angle) of the load or generator.
+        Default is 0.95.
+
+    Returns
+    -------
+    float or array-like
+        Voltage difference in kV.
+    """
+    sin_phi = np.sqrt(1 - cos_phi**2)
+    # Calculate the voltage difference using the formula from VDE-AR-N 4105
+    voltage_diff = np.abs(
+        (s_max * 1e6 / (v_nom * 1e3)) * (r_total * cos_phi + sign * x_total * sin_phi)
+    )  # in V
+    return voltage_diff / 1e3  # Convert to kV
+
+
+def voltage_diff_pu(
+    R_per_km: float | np.ndarray,
+    L_per_km: float | np.ndarray,
+    length: float,
+    num_parallel: int,
+    v_nom: float | np.ndarray,
+    s_max: float | np.ndarray,
+    cos_phi: float = 0.95,
+    sign: int = -1,
+) -> float | np.ndarray:
+    """
+    Calculate the voltage difference per unit of nominal voltage.
+
+    Parameters
+    ----------
+    R_per_km : float or array-like
+        Resistance per kilometer of the cable in ohm/km.
+    L_per_km : float or array-like
+        Inductance per kilometer of the cable in mH/km.
+    length : float
+        Length of the cable in km.
+    num_parallel : int
+        Number of parallel cables.
+    v_nom : int
+        Nominal voltage in kV.
+    s_max : float
+        Apparent power the cable must carry in MVA.
+    cos_phi : float, optional
+        Cosine phi of the load or generator. Default: 0.95.
+    sign : int, optional
+        Sign of the reactance. -1 for inductive and +1 for capacitive. Default is -1.
+
+    Returns
+    -------
+    float
+        Voltage difference in per unit of nominal voltage.
+    """
+    # Calculate total resistance and reactance for the given length and
+    # number of parallel cables
+    r_total = calculate_line_resistance(R_per_km, length, num_parallel)
+    x_total = calculate_line_reactance(L_per_km, length, num_parallel)
+
+    # Calculate the voltage drop or increase
+    delta_v = calculate_voltage_diff_per_line(
+        s_max, r_total, x_total, v_nom, sign=sign, cos_phi=cos_phi
+    )
+
+    # Convert voltage difference to per unit of nominal voltage
+    voltage_difference_pu = delta_v / v_nom
+
+    return voltage_difference_pu
+
+
+def select_cable(
+    edisgo_obj: EDisGo,
+    level: str,
+    apparent_power: float,
+    length: float = 0,
+    max_voltage_diff: float | None = None,
+    max_cables: int = 7,
+    cos_phi: float | None = 0.95,
```

Comment:
Please rename this parameter to power_factor to have it consistent with the rest of the code base. Also, I would prefer that there is no default value but that the default value is obtained from the configs. For this, the type of the component to be connected needs to be specified as input parameter.